src/EditorFeatures/CSharpTest/ConvertForToForEach/ConvertForToForEachTests.cs

```diff
@@ -0,0 +1,715 @@
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForToForEach)]
+        public async Task TestDifferentIndexerAndEnumeratorType()
+        {
+            await TestInRegularAndScriptAsync(
+@"using System;
+
+class MyList
+{
+  public string this[int i] { get; }
+
+  public Enumerator GetEnumerator() { }
+
+  public struct Enumerator { public object Current { get; } }
+}
+
+class C
+{
+    void Test(MyList list)
+    {
+        // need to use 'string' here to preserve original index semantics.
+        [||]for (int i = 0; i < list.Length; i++)
+        {
+            Console.WriteLine(list[i]);
+        }
+    }
+}",
+@"using System;
+
+class MyList
+{
+  public string this[int i] { get; }
+
+  public Enumerator GetEnumerator() { }
+
+  public struct Enumerator { public object Current { get; } }
+}
+
+class C
+{
+    void Test(MyList list)
+    {
+        // need to use 'string' here to preserve original index semantics.
+        foreach (string {|Rename:v|} in list)
+        {
+            Console.WriteLine(v);
+        }
+    }
+}", options: ImplicitTypeEverywhere());
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForToForEach)]
+        public async Task TestSameIndexerAndEnumeratorType()
+        {
+            await TestInRegularAndScriptAsync(
+@"using System;
+
+class MyList
+{
+  public object this[int i] { get; }
+
+  public Enumerator GetEnumerator() { }
+
+  public struct Enumerator { public object Current { get; } }
+}
+
+class C
+{
+    void Test(MyList list)
+    {
+        // can use 'var' here since hte type stayed the same.
+        [||]for (int i = 0; i < list.Length; i++)
+        {
+            Console.WriteLine(list[i]);
+        }
+    }
+}",
+@"using System;
+
+class MyList
+{
+  public object this[int i] { get; }
+
+  public Enumerator GetEnumerator() { }
+
+  public struct Enumerator { public object Current { get; } }
+}
+
+class C
+{
+    void Test(MyList list)
+    {
+        // can use 'var' here since hte type stayed the same.
+        foreach (var {|Rename:v|} in list)
+        {
+            Console.WriteLine(v);
+        }
+    }
+}", options: ImplicitTypeEverywhere());
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForToForEach)]
+        public async Task TestTrivia()
+        {
+            await TestInRegularAndScript1Async(
+@"using System;
+
+class C
+{
+    void Test(string[] array)
+    {
+        // trivia 1
+        [||]for /*trivia 2*/ ( /*trivia 3*/ int i = 0; i < array.Length; i++) /*trivia 4*/
+        // trivia 5
+        {
+            Console.WriteLine(array[i]);
+        } // trivia 6
+    }
+}",
+@"using System;
+
+class C
+{
+    void Test(string[] array)
+    {
+        // trivia 1
+        foreach /*trivia 2*/ ( /*trivia 3*/ string {|Rename:v|} in array) /*trivia 4*/
+        // trivia 5
+        {
+            Console.WriteLine(v);
+        } // trivia 6
+    }
+}");
+        }
+    }
+}
```

Comment:
Worth testing with a deconstruction, although I expect no refactoring will be offered.
`for (var (i, j) = (0, 0); ....)...`

Reply:
sure.

Reply:
#resolved